lib/data/external/AwsClient.js

```diff
@@ -117,18 +124,25 @@ class AwsClient {
         return callback(null, stream);
     }
     delete(objectGetInfo, reqUids, callback) {
-        // for backwards compatibility
-        const key = typeof objectGetInfo === 'string' ? objectGetInfo :
-          objectGetInfo.key;
+        const { key, dataStoreVersionId, deleteVersion } = objectGetInfo;
+        const log = createLogger(reqUids);
         const params = {
             Bucket: this._awsBucketName,
             Key: key,
         };
+        if (deleteVersion) {
+            params.VersionId = dataStoreVersionId;
+        }
         return this._client.deleteObject(params, err => {
             if (err) {
-                const log = createLogger(reqUids);
                 logHelper(log, 'error', 'error deleting object from ' +
                 'datastore', err, this._dataStoreName);
+                if (err.code === 'NoSuchVersion') {
+                    // data may have been deleted directly from the AWS backend
+                    // don't want to retry the delete and errors are not
+                    // sent back to client anyway, so no need to return err
+                    return callback();
```

Comment:
i don't follow the logic on why not return an internal error in this case?

Reply:
We can return the error, but it will just result in the delete being retried. The error doesn't get sent back to the user because at that point we have already deleted the metadata for the object:
https://github.com/scality/S3/blob/master/lib/data/wrapper.js#L203

And since we can expect this error to occur if someone deletes the object directly from the AWS backend, I didn't think it was necessary to retry the delete a few times since retries will not address that problem. But maybe we can just in case, if you think it's worthwhile.

Reply:
I'm not super familiar with the intricacies of versioning, but I'd say if that's the only case when the NoSuchVersion error is returned, then I like skipping the retries :)